src/App.jsx

```diff
@@ -5,25 +5,51 @@ import './App.css'
 import AddTodoForm from './AddTodoForm'
 import TodoList from './TodoList'
 
-function useSemiPersistentState() {
-  const [todoList, setTodoList] = useState(JSON.parse(localStorage.getItem('savedTodoList')) || []);
-  useEffect(() => {
-    localStorage.setItem('savedTodoList', JSON.stringify(todoList));
-  }, [todoList]);
-
-  return [todoList, setTodoList];
-};
 
 function App() {
-  const [todoList, setTodoList] = useSemiPersistentState();
+  const [todoList, setTodoList] = useState([]);
+  const [isLoading, setIsLoading] = useState(true);
+
+  const fetchData = () => {
+    return new Promise((resolve, reject) => {
+      setTimeout(() => {
+        const savedTodoList = localStorage.getItem('savedTodoList');
+        const parsedTodoList = savedTodoList ? JSON.parse(savedTodoList) : [];
+        resolve({ data: { todoList: parsedTodoList } });
+      }, 2000);
+    });
+  };
+
+  useEffect(() => {
+    fetchData()
+      .then((response) => {
+        setTodoList(response.data.todoList);
+        setIsLoading(false);
+        return response.data;
+      })
+    .then(result => {
+      setTodoList(result.todoList);
+      setIsLoading(false);
+    })
+    .catch(error => {
+      console.error('Error:', error);
+      setIsLoading(false);
+    });
```

Comment:
You could refactor this code using the `finally` block at the end of the then catch methods:

```
fetchData()
    .then((response) => {
      setTodoList(response.data.todoList);
      return response.data;
    })
    .catch(error => {
      console.error('Error:', error);
    })
    .finally(() => {
      setIsLoading(false);
    });
```
    
The finally() method of Promise instances schedules a function to be called when the promise is settled (either fulfilled or rejected).